Eventually a removed flow is recreated when Kytos is under high usage (possible race condition)

Original issue opened by @italovalcy at https://github.com/kytos/flow_manager/issues/128.

Hi folks,

The consistency check routine eventually recreates a flow when Kytos is under high usage (many requests in a short period of time), which indicates a possible race condition.

More specifically, the following end-to-end tests sometimes pass and sometimes fail:
```
_______________________________ TestE2EMefEline.test_070_repeat_creation_and_removal_of_ten_circuits_ten_times ________________________________

self = <tests.test_e2e_10_mef_eline.TestE2EMefEline instance at 0x7f757b5bd5f0>

    def test_070_repeat_creation_and_removal_of_ten_circuits_ten_times(self):
        for x in range(0, 10):
            evcs = {}
            for i in range(400, 410):
                payload = {
                    "name": "evc_%s" % i,
                    "enabled": True,
                    "dynamic_backup_path": True,
                    "uni_a": {
                        "interface_id": "00:00:00:00:00:00:00:01:1",
                        "tag": {"tag_type": 1, "value": i}
                    },
                    "uni_z": {
                        "interface_id": "00:00:00:00:00:00:00:02:1",
                        "tag": {"tag_type": 1, "value": i}
                    }
                }
                api_url = KYTOS_API + '/mef_eline/v2/evc/'
                response = requests.post(api_url, data=json.dumps(payload), headers={'Content-type': 'application/json'})
                assert response.status_code == 201
                data = response.json()
                assert 'circuit_id' in data
                evcs[i] = data['circuit_id']

            time.sleep(10)

            # make sure the evcs are active and the flows were created
            s1, s2 = self.net.net.get('s1', 's2')
            flows_s1 = s1.dpctl('dump-flows')
            flows_s2 = s2.dpctl('dump-flows')
            for vid in evcs:
                evc_id = evcs[vid]
                api_url = KYTOS_API + '/mef_eline/v2/evc/' + evc_id
                response = requests.get(api_url)
                assert response.status_code == 200
                evc = response.json()
                #should be active
                assert evc["active"] is True
                # search for the vlan id
                assert "dl_vlan=%s" % vid in flows_s1
                assert "dl_vlan=%s" % vid in flows_s2
                # search for the cookie, should have two flows
                assert len(re.findall(evc['id'], flows_s1, flags=re.IGNORECASE)) == 2
                assert len(re.findall(evc['id'], flows_s2, flags=re.IGNORECASE)) == 2

            # Delete the circuits
            for vid in evcs:
                evc_id = evcs[vid]
                api_url = KYTOS_API + '/mef_eline/v2/evc/' + evc_id
                response = requests.delete(api_url)
                assert response.status_code == 200

            time.sleep(10)

            # make sure the circuits were deleted
            api_url = KYTOS_API + '/mef_eline/v2/evc/'
            response = requests.get(api_url)
            assert response.status_code == 200
            assert response.json() == {}
            flows_s1 = s1.dpctl('dump-flows')
            flows_s2 = s2.dpctl('dump-flows')
>           assert len(flows_s1.split('\r\n ')) == 1, "round=%d - should have only 1 flow but had: \n%s" % (x, flows_s1)
E           AssertionError: round=8 - should have only 1 flow but had:
E              cookie=0xaa28bc57b4ba8f4a, duration=8.110s, table=0, n_packets=0, n_bytes=0, in_port="s1-eth1",dl_vlan=405 actions=strip_vlan,mod_vlan_vid:3461,output:"s1-eth3"
E              cookie=0xaa28bc57b4ba8f4a, duration=8.079s, table=0, n_packets=0, n_bytes=0, in_port="s1-eth3",dl_vlan=3461 actions=strip_vlan,mod_vlan_vid:405,output:"s1-eth1"
E              cookie=0x0, duration=220.814s, table=0, n_packets=144, n_bytes=6048, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535
E
E           assert 3 == 1
E            +  where 3 = len([' cookie=0xaa28bc57b4ba8f4a, duration=8.110s, table=0, n_packets=0, n_bytes=0, in_port="s1-eth1",dl_vlan=405 actions=...20.814s, table=0, n_packets=144, n_bytes=6048, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535\r\n'])
E            +    where [' cookie=0xaa28bc57b4ba8f4a, duration=8.110s, table=0, n_packets=0, n_bytes=0, in_port="s1-eth1",dl_vlan=405 actions=...20.814s, table=0, n_packets=144, n_bytes=6048, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535\r\n'] = <built-in method split of str object at 0x564093f85fd0>('\r\n ')
E            +      where <built-in method split of str object at 0x564093f85fd0> = ' cookie=0xaa28bc57b4ba8f4a, duration=8.110s, table=0, n_packets=0, n_bytes=0, in_port="s1-eth1",dl_vlan=405 actions=s...220.814s, table=0, n_packets=144, n_bytes=6048, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535\r\n'.split
```

As noted, the test basically creates 10 EVCs, waits a couple of seconds, makes sure the flows were created, removes the 10 evcs, waits a couple of seconds, makes sure the flows were removed, and so on. We repeat this procedure 10 times.

This test was run using the AmLight Kytos image (https://hub.docker.com/r/amlight/kytos), which is basically a master branch daily build image, with some PRs applied (e.g., #117 and more: https://github.com/amlight/kytos-docker/tree/master/patches). Also, we change the default STATS_INTERVAL setting to 3s and leave the CONSISTENCY_INTERVAL=0 (which means the consistency_check will run every 3s). The LINK_UP_TIMER is also changed to 1s.

This problem seems to be a race-condition situation because while the consistency check is running, other requests to remove EVCs are received for different switches and each thread seems to change the self.stored_flows in a uncontrolled manner (the shared resource).

I've enabled some logging and this is how it looks like (logs for the above test, cookie `0xaa28bc57b4ba8f4a  == 12261257070396477258 ==> evc ID 28bc57b4ba8f4a`):
```
May 11 01:35:28 da6a340a8416 kytos.napps.kytos/mef_eline:INFO main:246:  Removing EVC(28bc57b4ba8f4a, evc_405)

May 11 01:35:28 da6a340a8416 kytos.napps.kytos/flow_manager:INFO main:375:  _install_flows switches=[Switch('00:00:00:00:00:00:00:01')] command=delete flows={'flows': [{'cookie': 12261257070396477258, 'cookie_mask': 18446744073709551615}]}

May 11 01:35:29 da6a340a8416 kytos.napps.kytos/flow_manager:INFO main:375:  _install_flows switches=[Switch('00:00:00:00:00:00:00:02')] command=delete flows={'flows': [{'cookie': 12261257070396477258, 'cookie_mask': 18446744073709551615}]}

May 11 01:35:31 da6a340a8416 kytos.napps.kytos/flow_manager:INFO main:375:  _install_flows switches=[Switch('00:00:00:00:00:00:00:01')] command=add flows={'flows': [{'match': {'in_port': 1, 'dl_vlan': 405}, 'cookie': 12261257070396477258, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'action_type': 'set_vlan', 'vlan_id': 3461}, {'action_type': 'output', 'port': 3}]}]}

May 11 01:35:31 da6a340a8416 kytos.napps.kytos/flow_manager:INFO main:375:  _install_flows switches=[Switch('00:00:00:00:00:00:00:01')] command=add flows={'flows': [{'match': {'in_port': 3, 'dl_vlan': 3461}, 'cookie': 12261257070396477258, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 'c'}, {'action_type': 'set_vlan', 'vlan_id': 405}, {'action_type': 'output', 'port': 1}]}]}
```

In between those events, I could see that the function _store_changed_flows() sometimes overwrite a content previously changed. Attached the full log.
[syslog-2021-05-10.log.gz](https://github.com/kytos/flow_manager/files/6456054/syslog-2021-05-10.log.gz)

--- a/main.py
+++ b/main.py
@@ -1,6 +1,7 @@
 """kytos/flow_manager NApp installs, lists and deletes switch flows."""
 from collections import OrderedDict
 from copy import deepcopy
+from threading import Lock
 
 from flask import jsonify, request
 from pyof.foundation.base import UBIntBase
@@ -89,6 +90,8 @@ class Main(KytosNApp):
 
         # Storehouse client to save and restore flow data:
         self.storehouse = StoreHouse(self.controller)
+
+        self._storehouse_lock = Lock()
 
         # Format of stored flow data:
         # {'flow_persistence': {'dpid_str': {'flow_list': [
@@ -193,14 +196,14 @@ class Main(KytosNApp):
                 if command == 'add':
                     log.info('A consistency problem was detected in '
                              f'switch {dpid}.')
-                    self._install_flows(command, flow, [switch])
+                    self._install_flows(command, flow, [switch], save=False)
                     log.info(f'Flow forwarded to switch {dpid} to be '
                              'installed.')
             elif command == 'delete':
                 log.info('A consistency problem was detected in '
                          f'switch {dpid}.')
                 command = 'delete_strict'
-                self._install_flows(command, flow, [switch])
+                self._install_flows(command, flow, [switch], save=False)
                 log.info(f'Flow forwarded to switch {dpid} to be deleted.')
 
     def check_storehouse_consistency(self, switch):
@@ -218,7 +221,7 @@ class Main(KytosNApp):
                          f'switch {dpid}.')
                 flow = {'flows': [installed_flow.as_dict()]}
                 command = 'delete_strict'
-                self._install_flows(command, flow, [switch])
+                self._install_flows(command, flow, [switch], save=False)
                 log.info(f'Flow forwarded to switch {dpid} to be deleted.')
             else:
                 serializer = FlowFactory.get_class(switch)
@@ -232,7 +235,7 @@ class Main(KytosNApp):
                              f'switch {dpid}.')
                     flow = {'flows': [installed_flow.as_dict()]}
                     command = 'delete_strict'
-                    self._install_flows(command, flow, [switch])
+                    self._install_flows(command, flow, [switch], save=False)
                     log.info(f'Flow forwarded to switch {dpid} to be deleted.')
 
     # pylint: disable=attribute-defined-outside-init
@@ -429,13 +432,14 @@ class Main(KytosNApp):
 
         return jsonify({"response": "FlowMod Messages Sent"})
 
-    def _install_flows(self, command, flows_dict, switches=[]):
+    def _install_flows(self, command, flows_dict, switches=[], save=True):
         """Execute all procedures to install flows in the switches.
 
         Args:
             command: Flow command to be installed
             flows_dict: Dictionary with flows to be installed in the switches.
             switches: A list of switches
+            save: A boolean to save flows in the storehouse (True) or not
         """
         for switch in switches:
             serializer = FlowFactory.get_class(switch)
@@ -454,7 +458,9 @@ class Main(KytosNApp):
                 self._add_flow_mod_sent(flow_mod.header.xid, flow, command)
 
                 self._send_napp_event(switch, flow, command)
-                self._store_changed_flows(command, flow_dict, switch)
+                if save:
+                    with self._storehouse_lock:
+                        self._store_changed_flows(command, flow_dict, switch)
 
     def _add_flow_mod_sent(self, xid, flow, command):
         """Add the flow mod to the list of flow mods sent."""

